src/blockchain/chain.rs

```diff
@@ -107,41 +94,28 @@ where
 
     /// Handle a block whose ancestors are on disk.
     fn handle_connected_block(&mut self, block_hash: B::BlockHash, block: B::Block) {
-        use chain_core::property::{Block, LeaderSelection, Ledger, Settings};
-
-        let current_tip = self.state.settings.read().unwrap().tip();
+        let current_tip = self.state.tip();
 
-        // Quick optimization: don't do anything if the incoming block
-        // is already the tip. Ideally we would bail out if the
-        // incoming block is on the tip chain, but there is no quick
-        // way to check that.
-        if block_hash != current_tip {
-            if current_tip == block.parent_id() {
-                // FIXME: deadlock risk. But this issue will go away
-                // when we merge the states.
-                let mut ledger = self.state.ledger.write().unwrap();
-                let mut settings = self.state.settings.write().unwrap();
-
-                let leadership_diff = self.state.leaders.diff(&block).unwrap();
-                let ledger_diff = { ledger.diff(block.transactions()).unwrap() };
-                let setting_diff = settings.diff(&block).unwrap();
-
-                self.state.leaders.apply(leadership_diff.clone()).unwrap();
-                ledger.apply(ledger_diff.clone()).unwrap();
-                settings.apply(setting_diff.clone()).unwrap();
-
-                self.change_log
-                    .push((leadership_diff, ledger_diff, setting_diff));
+        match self.state.apply_block(&block.header(), block.messages()) {
+            Ok(state) => {
                 self.storage.put_block(&block).unwrap();
-                self.storage
-                    .put_tag(LOCAL_BLOCKCHAIN_TIP_TAG, &block_hash)
-                    .unwrap();
-            } else {
-                // TODO chain state restoration ?
-                unimplemented!()
+                if block_hash != current_tip {
+                    self.storage
+                        .put_tag(LOCAL_BLOCKCHAIN_TIP_TAG, &block_hash)
+                        .unwrap();
+                }
+
+                // TODO: update with the multiverse here
+                self.state = state;
             }
+            Err(error) => error!("Error with incoming block: {}", error),
         }
     }
+
+    /// return the current tip hash and date
+    pub fn get_tip(&self) -> B::BlockHash {
```

Comment:
How about gradually removing `get_` from method names so that the APIs are more conventional for Rust?

Reply:
agree. This PR is a bit massive and there are lot of component that are a bit hacky. I'd like to merge this ASAP and then we can start the surgery and do the proper cleanup of jormungandr.